pkg/api/validation/validation.go

```diff
@@ -1938,6 +1938,42 @@ func ValidateSecretUpdate(newSecret, oldSecret *api.Secret) field.ErrorList {
 	return allErrs
 }
 
+// ValidateConfigMapName can be used to check whether the given ConfigMap name is valid.
+// Prefix indicates this name will be used as part of generation, in which case
+// trailing dashes are allowed.
+func ValidateConfigMapName(name string, prefix bool) (bool, string) {
+	return NameIsDNSSubdomain(name, prefix)
+}
+
+// ValidateConfigMap tests whether required fields in the ConfigMap are set.
+func ValidateConfigMap(cfg *api.ConfigMap) field.ErrorList {
+	allErrs := field.ErrorList{}
+	allErrs = append(allErrs, ValidateObjectMeta(&cfg.ObjectMeta, true, ValidateConfigMapName, field.NewPath("metadata"))...)
+
+	totalSize := 0
+
+	for key, value := range cfg.Data {
+		if !IsSecretKey(key) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("data").Key(key), key, fmt.Sprintf("must have at most %d characters and match regex %s", validation.DNS1123SubdomainMaxLength, SecretKeyFmt)))
+		}
+		totalSize += len(value)
```

Comment:
We should count both keys and data towards the total, yes.

On Thu, Jan 21, 2016 at 2:55 PM, Paul Morie <notifications@github.com>
wrote:

> In pkg/api/validation/validation.go
> <https://github.com/kubernetes/kubernetes/pull/19909#discussion_r50453918>
> :
>
> > +func ValidateConfigMapName(name string, prefix bool) (bool, string) {
> > +	return NameIsDNSSubdomain(name, prefix)
> > +}
> > +
> > +// ValidateConfigMap tests whether required fields in the ConfigMap are set.
> > +func ValidateConfigMap(cfg *api.ConfigMap) field.ErrorList {
> > +	allErrs := field.ErrorList{}
> > +	allErrs = append(allErrs, ValidateObjectMeta(&cfg.ObjectMeta, true, ValidateConfigMapName, field.NewPath("metadata"))...)
> > +
> > +	totalSize := 0
> > +
> > +	for key, value := range cfg.Data {
> > +		if !IsSecretKey(key) {
> > +			allErrs = append(allErrs, field.Invalid(field.NewPath("data").Key(key), key, fmt.Sprintf("must have at most %d characters and match regex %s", validation.DNS1123SubdomainMaxLength, SecretKeyFmt)))
> > +		}
> > +		totalSize += len(value)
>
> @thockin <https://github.com/thockin> I'm wondering if we need to count
> the size as keys and data, since the size limit of 1MB comes in part from
> the limit of what is recommended to go into an etcd record.
>
> cc @smarterclayton <https://github.com/smarterclayton>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19909/files#r50453918>.
>